baybe/campaign.py

```
@@ -371,21 +437,87 @@ def recommend(
             self._measurements_exp.fillna({"FitNr": self.n_fits_done}, inplace=True)
 
         # Prepare the search space according to the current campaign state
-        annotated_searchspace = evolve(
-            self.searchspace,
-            discrete=AnnotatedSubspaceDiscrete.from_subspace(
-                self.searchspace.discrete, self._searchspace_metadata
-            ),
-        )
+        if self.searchspace.type is SearchSpaceType.DISCRETE:
+            # TODO: This implementation should at some point be hidden behind an
+            #   appropriate public interface, like `SubspaceDiscrete.filter()`
+            mask_todrop = self._searchspace_metadata[_EXCLUDED].copy()
+            if not self.allow_recommending_already_recommended:
+                mask_todrop |= self._searchspace_metadata[_RECOMMENDED]
+            if not self.allow_recommending_already_measured:
+                mask_todrop |= self._searchspace_metadata[_MEASURED]
+            if (
+                not self.allow_recommending_pending_experiments
+                and pending_experiments is not None
+            ):
+                mask_todrop |= pd.merge(
+                    self.searchspace.discrete.exp_rep,
+                    pending_experiments,
+                    indicator=True,
+                    how="left",
+                )["_merge"].eq("both")
+            searchspace = evolve(
+                self.searchspace,
+                discrete=FilteredSubspaceDiscrete.from_subspace(
+                    self.searchspace.discrete, ~mask_todrop.to_numpy()
+                ),
+            )
+        else:
+            searchspace = self.searchspace
+
+        # Pending experiments should not be passed to non-predictive recommenders
+        # to avoid complaints about unused arguments, so we need to know of what
```

I think there was supposed to be a warning here in campaign when pending expts are passed to non-pred recommenders

---

Wait what!? We that entire suuuuper long discussion [here](https://github.com/emdgroup/baybe/pull/423#discussion_r1851622538) where we all converged that:
* Recommenders should throw an error
* And the campaign needs to take that this situation doesn't occur in the first place

In fact, that proposal came from [you](https://github.com/emdgroup/baybe/pull/423#discussion_r1875580106). And that was one of the big motivations for #457, because the campaign needs to know what recommender will be in place next.

---

ah I see its in non-pre, probably confused that

---

puhhh 😄 days of work almost wasted